Dockerfile and fix rspec

directionless opened the pull request:
This creates a Dockerfile suitable both for running tests, as well as creating releasable containers forgeneral use. It does this by leveraging docker's buildkit and build arg functionality. There is some usage documentation in `/docs/docker.rst` and in the `Dockerfile` itself.

To support this, the rpsec tests have been updated a bit. They now run their dependency detection in the shared files, and have some additional case statements to emit more clear and succinct warnings.

Additionally, the travis-ci configs have been updated to remove broken ruby platforms. This is done by (a) shifting away from `osx` to the more explicit `xcode11.3` and `xcode9.4`, and disabling the broken `1.9.3`.

Fixes #1682, fixes #1453, fixes #1437, fixes #1357, fixes #992, closes #1453

jordansissel commented:
Unrelated to this PR, several rspec tests tend to always fail.  I'll keep an eye on this PR in the meantime. In general, if you're looking for what   a "successful" rspec run looks like these days, it'll be that most tests pass but maybe 10-20 fail. I've been slowly working on disabling the broken test suites but haven't chaned much recently.

Thank you for this work!

jordansissel commented:
Thinking more on it, I bet we can use this work to help find common failing rspecs and either disable them (as a separate PR that I or someone can do) or fix them. It'd be nice to have this incentive to get the test suite green again :)

directionless commented:
I started going through the tests, and I discovered that they're in great shape.

---

I updated a bunch of the rspec skip mechanisms, now it'll display a fairly succinct message. Other idioms are to use `describe Class, if: HAVE...` but that pattern didn't provide useful output about missing something. Adding a skip call to a `before do` works, but spams output for every skipped test. So, I tried this. What do you think?

---

The docker stuff got complicated. I wanted to:
* support a minimalish build, since the python/npm/perl stuff makes for much bigger installs
* support easy testing, as well as pushing a reasonable final docker docker
* Only run the desired steps.

The newish docker bjuildkit stuff allows this.  It uses 2 args to specify what should happen, and buildkit then resolves the build chain. It's kinda neat. 

I've added some make targets to help.

---

Still do to:
 * [ ] Update docs
 * [ ] fix or disable python tests

directionless commented:
Python tests are fixed!

As a side note, debugging is hard, since most of the tool stdout is suppressed.

directionless commented:
I've been running into bunch of issues around travis and ruby. There's some mention of it around the various other issues:
* https://github.com/travis-ci/travis-ci/issues/7848
* https://github.com/travis-ci/travis-ci/issues/2793
* https://github.com/rvm/rvm/issues/3923

directionless commented:
The travis-ci ruby 1.9.3 seems broken. So I removed it.
```
$ source ${HOME}/.venv/bin/activate && PATH=${PATH}:${HOME}/.local/bin bundle exec rspec
/home/travis/.rvm/gems/ruby-1.9.3-p551/gems/backports-3.15.0/lib/backports/std_lib.rb:9:in `require': /home/travis/.rvm/gems/ruby-1.9.3-p551/gems/json-2.3.0/lib/json/common.rb:156: syntax error, unexpected tPOW, expecting ')' (SyntaxError)
    Parser.new(source, **(opts||{})).parse
                         ^
/home/travis/.rvm/gems/ruby-1.9.3-p551/gems/json-2.3.0/lib/json/common.rb:156: syntax error, unexpected ')', expecting keyword_end
    Parser.new(source, **(opts||{})).parse
                                    ^
/home/travis/.rvm/gems/ruby-1.9.3-p551/gems/json-2.3.0/lib/json/common.rb:179: syntax error, unexpected tPOW, expecting ')'
    Parser.new(source, **(opts||{})).parse
                         ^
/home/travis/.rvm/gems/ruby-1.9.3-p551/gems/json-2.3.0/lib/json/common.rb:179: syntax error, unexpected ')', expecting keyword_end
    Parser.new(source, **(opts||{})).parse
                                    ^
	from /home/travis/.rvm/gems/ruby-1.9.3-p551/gems/backports-3.15.0/lib/backports/std_lib.rb:9:in `require_with_backports'
```

directionless commented:
Hi @jordansissel, I believe this is ready for review. 

I don't know if you'll like how I updated rspec. This works, but there may be better idioms.

This PR ends up doing several things. I'm happy to split it up if you'd rather

directionless commented:
I wanted to check in on this PR. Thought?

abitrolly commented:
@directionless if test fix could be split into separate PR, I am sure it will speed up the review process.

jordansissel commented:
This PR is still on my list to review. The specs have gone through a few changes which now cause conflicts on this branch.  That said, I will try to review soon.

jordansissel commented:
I quickly skimmed over the changes in this PR. I believe I will be able to merge it once I fix the conflicts (which shouldn't be too hard).

I like this change as it will enable others (and me!!) to more easily run the full test suite without having to worry too much about what tools are installed in the dev environment. Thank you!

abitrolly commented:
@jordansissel I usually use https://github.com/mhagger/git-imerge for rebasing hard merges.

```
git-imerge rebase master
```
```
while not done:
    git-imerge diagram
    <fix the conflict that is presented to you>
    <"git add" the files that you changed>
    git-imerge continue
```
```
git-imerge finish
```

jordansissel commented on Dockerfile:
```diff
@@ -1,16 +1,86 @@
+RUN gem install --no-ri --no-rdoc bundler ffi
+RUN mkdir /origsrc
+COPY . /origsrc
+RUN cd /origsrc && bundle install
```
This step fails for me. I'm not sure what's causing it since bundler is rather _obviously_ installed in the previous step.

```
 => CACHED [test 2/6] RUN apt-get update  && apt-get install --no-install  0.0s
 => CACHED [test 3/6] RUN gem install --no-ri --no-rdoc bundler ffi        0.0s
 => CACHED [test 4/6] RUN mkdir /origsrc                                   0.0s
 => [test 5/6] COPY . /origsrc                                             0.1s
 => ERROR [test 6/6] RUN cd /origsrc && bundle install                     0.5s
------
 > [test 6/6] RUN cd /origsrc && bundle install:
#13 0.471 /usr/lib/ruby/2.5.0/rubygems.rb:289:in `find_spec_for_exe': can't find gem bundler (>= 0.a) with executable bundle (Gem::GemNotFoundException)
#13 0.471 	from /usr/lib/ruby/2.5.0/rubygems.rb:308:in `activate_bin_path'
#13 0.471 	from /usr/local/bin/bundle:23:in `<main>'
------
executor failed running [/bin/sh -c cd /origsrc && bundle install]: exit code: 1
make: *** [Makefile:61: .docker-test-everything] Error 1
```

jordansissel replied:
If I make it list the bundler version before trying to run it, it does show that bundler is installed, and yet, `bundle install` itself fails inside of bundler? somehow? I've never seen this behavior before.
```
#14 [test 7/7] RUN cd /origsrc; gem list bundler;  bundle install
#14 sha256:152d7cbdd81380115de7ebb36522f118b1b0095aa4b642ab2c803eb3e3a03760
#14 0.689 bundler (2.2.25)
#14 0.748 /usr/lib/ruby/2.5.0/rubygems.rb:289:in `find_spec_for_exe': can't find gem bundler (>= 0.a) with executable bundle (Gem::GemNotFoundException)
```

jordansissel replied:
Oh, I got it sorted. It's _extremely_ confusing due to the error message. This blog post explains it: https://bundler.io/blog/2019/05/14/solutions-for-cant-find-gem-bundler-with-executable-bundle.html

I'll work on a fix.

directionless commented:
Sorry I haven't had a chance to come back to this. Work moved on and all...

Is there something I can help with? I know the docker stuff is a bear

abitrolly commented:
@directionless https://github.com/jordansissel/fpm/pull/1818#issuecomment-899894354

jordansissel commented:
https://github.com/jordansissel/fpm/pull/1818 merged with changes credited to you from this PR. Thank you!

jordansissel commented:
Closing -- this pr was merged as separately #1818
